lib/devices.js

```diff
@@ -0,0 +1,77 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+module.exports = function (log, db, push) {
+  return {
+    upsert: upsert,
+    synthesizeName: synthesizeName
+  }
+
+  function upsert (request, sessionToken, deviceInfo) {
+    var operation, event, result
+    if (deviceInfo.id) {
+      operation = 'updateDevice'
+      event = 'device.updated'
+    } else {
+      operation = 'createDevice'
+      event = 'device.created'
+    }
+    var isPlaceholderDevice = Object.keys(deviceInfo).length === 0
+
+    return db[operation](sessionToken.uid, sessionToken.tokenId, deviceInfo)
+      .then(function (device) {
+        result = device
+        return log.activityEvent(event, request, {
+          uid: sessionToken.uid.toString('hex'),
+          device_id: result.id.toString('hex'),
+          is_placeholder: isPlaceholderDevice
+        })
+      })
+      .then(function () {
+        if (operation === 'createDevice') {
+          push.notifyDeviceConnected(sessionToken.uid, result.name, result.id.toString('hex'))
+          return log.event('device:create', request, {
+            uid: sessionToken.uid,
+            id: result.id,
+            type: result.type,
+            timestamp: result.createdAt,
+            isPlaceholder: isPlaceholderDevice
```

Comment:
Since we're now creating device records implicity as part of `/account/create and /account/login`, will `isPlaceholder` always be `true` in practice.  ISTM that we'll always create a placeholder record before the device has a chance to register itself.